### PR TITLE
feat: add support for computed members

### PR DESCRIPTION
**Pull request by switchdk**
- [x] Use new IAM module
- [x] Update from Template
- [x] Fix documentation to use correct values from code
- [x] Align code with new IAM module
- [x] Align tests with code and validate outcome

### CODE REVIEW

File: README.tfdoc.hcl [hunk 1]
```diff
@@ -426,8 +425,8 @@ section {
       }
 
       variable "traffic" {
-        type        = list(traffic)
-        default     = []
+        type        = any
+        default     = null
```
Comment (waxb) [inline README.tfdoc.hcl:429]: please remove `default = null` from every tfdoc you see. btw why did you change this from list to `any`?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -7,8 +7,14 @@ on:
...
   pull_request:
 
+env:
+  GO_VERSION: 1.18.7
+  TERRAFORM_VERSION: 1.3.6
+  TERRAMATE_VERSION: 0.2.9
+  TFLINT_VERSION: v0.44.1
+
 concurrency:
-  group: terraform-google-cloud-run
```
Comment (waxb) [inline .github/workflows/tests.yml:11]: please run `make template/adjust`

File: README.tfdoc.hcl [hunk 3]
```diff
@@ -494,46 +486,10 @@ section {
           END
         }
 
-        attribute "generation" {
```
Comment (waxb) [inline README.tfdoc.hcl:497]: nice catch

File: README.tfdoc.hcl [hunk 4]
```diff
@@ -611,46 +560,11 @@ section {
             END
           }
 
-          attribute "generation" {
```
Comment (waxb) [inline README.tfdoc.hcl:614]: also nice catch

**Review by waxb — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 2 above)
Comment (switchdk) [reply .github/workflows/tests.yml:11]: hahaha, I forgot. NIce catch, thank you

File: README.tfdoc.hcl (hunk 1 above)
Comment (switchdk) [reply README.tfdoc.hcl:429]: I only updated it so it would reflect `variables.tf` but on second thought, it makes more sense to have the documentation state `list(traffic)`. I will remove `default = null` from the `README.tfdoc.hcl`. Thanks

File: test/unit-complete/main.tf [hunk 5]
```diff
@@ -66,99 +81,126 @@ module "test" {
...
+        "allUsers",
+        "allAuthenticatedUsers",
+        "user:user@${local.org_domain}",
+        "serviceAccount:service-account-name@${local.project_id}.iam.gserviceaccount.com",
+        "group:group@${local.org_domain}",
+        "domain:${local.org_domain}",
+        "projectOwner:${local.project_id}",
+        "projectEditor:${local.project_id}",
+        "projectViewer:${local.project_id}",
+        "computed:service-account-name@${local.project_id}.iam.gserviceaccount.com"
```
Comment (mariux) [inline test/unit-complete/main.tf:136]: please use example.com and a dummy project

Comment (mariux) [inline test/unit-complete/main.tf:136]: that's not how computed works... you need to use the key of the desired entry in the map you provide.

**Review by mariux — CHANGES_REQUESTED** [review]
please add a non computed reference in call that allow to do so to validate we can use computed values directly.

File: test/unit-complete/main.tf (hunk 5 above)
Comment (switchdk) [reply test/unit-complete/main.tf]: Alright. Noted. Thanks.

Comment (switchdk) [reply test/unit-complete/main.tf]: hahaha, I had a momentary lapse of reasoning when did the computed members. Thanks for spotting 🙏🏻

File: variables.tf [hunk 6]
```diff
@@ -56,31 +56,42 @@ variable "domain_mapping" {
...
   default     = null
 }
 
+variable "computed_members_map" {
+  type        = map(string)
+  description = "(Optional) A map of members to replace in `members` to handle terraform computed values. Will be ignored when policy bindings are used."
+  default     = {}
+
+  validation {
+    condition     = alltrue([for k, v in var.computed_members_map : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|principal|principalSet):)", v))])
```
Comment (waxb) [inline variables.tf:76]: please update the condition [accordingly](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_run_service_iam)

if you copy-paste, copy from the child iam module :-)

**Review by waxb — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 6 above)
Comment (switchdk) [reply variables.tf]: Thank you and nice catch. Not sure where I copied that from but apparently I didn't read it. Has been fixed.

File: examples/example/main.tf [hunk 7]
```diff
@@ -1,3 +1,9 @@
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# EXAMPLE FULL USAGE OF THE TERRAFORM-MODULE-TEMPLATE MODULE
+#
+# And some more meaningful information.
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (waxb) [inline examples/example/main.tf:6]: what is this? and why TERRAFORM-MODULE-TEMPLATE?

File: main.tf [hunk 8]
```diff
@@ -119,24 +111,20 @@ resource "google_cloud_run_service" "service" {
 resource "google_cloud_run_domain_mapping" "domain_mapping" {
   count = var.module_enabled && var.domain_mapping != null ? 1 : 0
 
-  name     = var.domain_mapping.name
-  location = var.domain_mapping.location
-  project  = var.project
+  name     = google_cloud_run_service.service[0].name
+  location = google_cloud_run_service.service[0].location
+  project  = google_cloud_run_service.service[0].project
```
Comment (waxb) [inline main.tf:116]: is this always will be `google_cloud_run_service`'s attributes? can't i map something different (as before from the variable)?

File: main.tf [hunk 9]
```diff
@@ -86,7 +82,7 @@ resource "google_cloud_run_service" "service" {
   }
 
   dynamic "traffic" {
-    for_each = var.traffic
+    for_each = var.traffic != null ? var.traffic : []
```
Comment (waxb) [inline main.tf:85]: why? `traffic` defaults to `[]`

File: README.tfdoc.hcl [hunk 10]
```diff
@@ -427,7 +426,6 @@ section {
 
       variable "traffic" {
         type        = list(traffic)
-        default     = []
```
Comment (waxb) [inline README.tfdoc.hcl:430]: why? it is defaulting to `[]`

File: variables.tf [hunk 11]
```diff
@@ -38,7 +38,7 @@ variable "autogenerate_revision_name" {
 variable "traffic" {
   description = "(Optional) A list of traffic specifies how to distribute traffic over a collection of Knative Revisions and Configurations Structure."
   type        = any
-  default     = []
+  default     = null
```
Comment (waxb) [inline variables.tf:41]: why?
